Compile with pip Nvidia packages...

PhilipVinc opened the pull request.

PhilipVinc commented:
Installing this branch with 

```python
pip install --no-build-isolation --verbose git+https://github.com/mpi4jax/mpi4jax.git@pv/test-install
```

works and correctly picks up the Nvidia packages if they are installed.

However one then needs to set the `LD_LIBRARY_PATH` because the path is not hardcoded... with something like
```python
export LD_LIBRARY_PATH=[...]/ENV_NAME/lib/python3.11/site-packages/nvidia/cuda_runtime/lib:$LD_LIBRARY_PATH
```

Nevertheless, this still gives raise to some annoying ptas errors..

PhilipVinc commented:
Though it works for standard code

PhilipVinc commented:
But this is nice. It allows to avoid requiring CUDA installed....

PhilipVinc commented:
This now sets the rpath so it bakes the compile-time libraries if it detects it's obtained from pip...

PhilipVinc commented:
@dionhaefner what can we do to package this properly?

dionhaefner commented on docs/installation.rst:
```diff
@@ -67,11 +67,23 @@ Installation with NVIDIA GPU support (CUDA)
 
-   To use JAX on the GPU, make sure that your ``jaxlib`` is `built with CUDA support <https://github.com/google/jax#installation>`_.
+   There are 2 ways to install jax with CUDA support:
+   - using a pypi-distributed CUDA installation (suggested by jax developers) ``pip install -U 'jax[cuda12_pip]' -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html`` 
```
Not `jax[cuda]`?

PhilipVinc replied:
`jax[cuda]` does not exist.
`jax[cuda12]` does exist, but won't work because of other bugs in mpi4jax / recent changes in XLA primitive registration (or at least, I could not get it to work).

If we fix those bugs I will update the guide.

dionhaefner replied:
Should we add a warning then not to use `jax[cuda12]`? This is the recommended mode of installing JAX after all (while `jax[cuda12_pip]` is deprecated per the docs).

PhilipVinc replied:
added relevant notices

dionhaefner commented:
This looks generally fine to me. 

Setting `rpath` means that mpi4jax may break violently if the user ever updates the CUDA libs but I guess it's the best we can do, barring using `dlopen` at runtime like XLA does.

I think the most important task now is to test all modes of installation on CI. That should give us enough confidence to ship this new feature pretty much as-is.

PhilipVinc commented:
@dionhaefner I added a test for the build infrastructure using PyPi distributed cuda.

I also added the relevant instructions and a mention that [cuda12] is not yet supported.

Anything else?

dionhaefner commented on .github/workflows/build-gpu-ext.yml:
```diff
@@ -61,6 +71,14 @@ jobs:
+    - name: Build GPU extensions
+      if: matrix.cuda == 'pypi'
       run: |
+        pip install mpi4py
```
Why do we need this? `mpi4py` should be installed since the last step already.

PhilipVinc commented:
Is it fine for you if I merge and tag a new release?

dionhaefner commented on docs/installation.rst:
```diff
@@ -67,11 +67,24 @@ Installation with NVIDIA GPU support (CUDA)
+   There are 3 ways to install jax with CUDA support:
+   - using a pypi-distributed CUDA installation (suggested by jax developers) ``pip install -U 'jax[cuda12_pip]' -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html`` 
+   - using the locally-installed CUDA version, which must be compatible with jax. ``pip install -U 'jax[cuda12_local]' -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html`` 
+   The procedure to install ``mpi4jax`` for the two situations is different.
```
We could recommend `--no-build-isolation` always for CUDA builds? Can't come up with a case where this would hurt to have.

dionhaefner commented:
Fine with me, I could pick some nits with the logic in `setup.py` and the wording in the docs, but feels like this is up for a refactor soon anyway once we have to support `jax[cuda12]` (when `jax[cuda12_pip]` goes away).
